test/test_queue_routes.py

```diff
@@ -3,83 +3,360 @@
+                     data=json.dumps([task_to_add]),
+                     content_type='application/json'
+                     )
+    assert rv.status_code == 200
+    yield add_task
 
 def test_get_main(client):
     """Test if we can get the home page."""
     assert client.get('/').status_code == 200
 
 
-def test_get_queue(client):
+def test_queue_get(client):
     """Test if we can get the queue."""
-
     rv = client.get('/mxcube/api/v0.1/queue')
     assert rv.status_code == 200
 
 
-def test_get_item(client):
+def test_queue_get_item(client):
     """Test if we the queue has the intial sample, only mockups hwobj."""
     rv = client.get('/mxcube/api/v0.1/queue')
-
     assert (rv.status_code == 200 and json.loads(rv.data).get('1:01'))
 
-
-def test_add_sample(client):
+def test_queue_add_item(client):
     """Test if we can add a sample."""
-    sample_to_add = {
-        'code': 'matr1_5',
-        'checked': True,
-        'sampleName': 'Sample-105',
-        'sampleID': '1:05',
-        'tasks': [],
-        'location': '1:5',
-        'defaultPrefix': 'local-user',
-        'type': 'Sample'
-    }
+    rv = client.post('/mxcube/api/v0.1/queue',
+                     data=json.dumps([test_sample]),
+                     content_type='application/json'
+                     )
+    assert rv.status_code == 200
+
+def test_add_and_get_sample(client, add_sample):
+    """Test if we can add a sample. The sample is added by a fixture."""
+    rv = client.get('/mxcube/api/v0.1/queue')
+    assert (rv.status_code == 200 and json.loads(rv.data).get('1:05'))
+
+def test_add_and_get_task(client, add_sample, add_task):
+    """Test if we can add a task to the sample."""
+    rv = client.get('/mxcube/api/v0.1/queue')
+    assert (rv.status_code == 200 and len(json.loads(rv.data).get('1:05')['tasks']) == 1)
+
+def test_add_and_edit_task(client, add_sample, add_task):
+    """Test if we can add edit a task i the sample in the queue."""
+    rv = client.get('/mxcube/api/v0.1/queue')
+    queue_id = json.loads(rv.data).get('1:05')['queueID']
+    task_queue_id = json.loads(rv.data).get('1:05')['tasks'][0]['queueID']
+    assert (rv.status_code == 200 and len(json.loads(rv.data).get('1:05')['tasks']) == 1)
+
+    task_to_update = copy.deepcopy(test_edit_task)
+    parameter_to_update = 'num_images'
+    parameter_update_value = 10
+    task_to_update['parameters'][parameter_to_update] = parameter_update_value
+    rv = client.post(('/mxcube/api/v0.1/queue/{}/{}').format(queue_id, task_queue_id),
+                     data=json.dumps(task_to_update),
+                     content_type='application/json'
+                     )
+    assert (rv.status_code == 200 and json.loads(rv.data).get('parameters')[parameter_to_update] == parameter_update_value)
+
+def test_queue_start(client, add_sample, add_task):
+    """Test if we can start the queue. The queue requires a sample and a task to start which are added by fixtures."""
```

Comment:
What do you think about also checking if the queue was annually started ?

Reply:
We had problems writing tests for the routes that are using gevent; pytest doesn't seem to follow the gevent calls; maybe because the calls are spawned on another thread. That is the reason we skipped pause and unpause route calls for now. The same applies for these. We couldn't understand completely how to cater this for now so we will have a look at it later.

Reply:
I don't see how the gevent calls could be spawned in another thread and even if they would be how would it be a problem ?. Your doing a HTTP call through the test client so that should in principle be independent on whatever thread running on the server, or am I wrong ?

What do you mean when you say that pytest have problems following gevent code, how did you do the actual queue state verification ?

Reply:
@13bscsaamjad are you talking about coverage report? You need to configure pytest with
concurrency=gevent. 

See https://stackoverflow.com/questions/22890779/determining-test-coverage-with-py-test-gevent

Reply:
@mguijarr , good to see you around!
we try to get the state of the queue by calling `/queue_state` (after starting it, tried with some delay as well) and then looking at the queueStatus field, it is alwasy queueStopped.
We follow the call until `HardwareObjects.QueueManager.execute()`, 
the call `self._root_task = gevent.spawn(self.__execute_task)` seems that is never executed, or well, it has no effect, the greenlet is created but we cannot see it running... we do not see any error as well.

Reply:
@meguiraun I am very busy with BLISS but sometimes I read messages here ;) I think I will see you at NOBUGS next week right ? I will talk about this error with Marcus tomorrow... For the call you mention I guess it returns (self._root_task == a gevent task) but if it is not running it means the gevent loop is not running...

Reply:
I see, so its likely that the gevent loop is not running properly. It might be sufficient to move the monkey patching done in the mxcube3-server startup script to mxcube3/__init__.py.

As it is now the test code is running un-patched so that is likely a problem. I'm however not sure that it's the problem that we are seeing here (but probable). I think it's quite important that we find out the reason for this issue to make sure that we are not doing something wrong.

Also I did not find a test that tests queue execution, I guess that would be problematic for the same reason ? Once solved I guess it would be good to have such a test.

Reply:
Applied moving the monkey patching and it worked for start, stop and abort queue tests.

For queue execution, the call is blocking and it expects something to be returned. However, the server does not return anything. I could not figure it out until now. I might add those tests later.